Unify OpenSearch startup connection gate: phase-aware retry/fallback across both startup paths

## Description

dotCMS has **inconsistent OpenSearch (OS) connection gating at startup**, and the OS readiness gate does not respect the migration phase. Three concrete gaps were confirmed against the current code (see file:line below). Together they cause an opaque, late-failing crash when OS is unreachable/misconfigured during startup, instead of a fast, actionable failure with phase-correct behavior (fallback during migration, abort once OS is primary).

This is the follow-up to a finding made on 2026-06-12 while smoke-testing Phase 3 against the `os-migration` limited-user stack: the real failure was a TLS scheme mismatch (`http://` against an OS 3.x port that only accepts `https`), surfacing as `ConnectionClosedException` deep inside `createContentIndex` ~30s into startup (`SystemExitManager - Startup failure`), with no actionable message.

### Current behavior (confirmed)

| Gate | Where | Today |
|---|---|---|
| ES readiness | `ESIndexAPI.waitUtilIndexReady()` (`ESIndexAPI.java:1150`) | Retries `ES_CONNECTION_ATTEMPTS` (default 24, 5s sleep); on exhaustion → `SystemExitManager.immediateExit(1)`. Always aborts. |
| OS readiness | `OSIndexAPIImpl.waitUtilIndexReady()` (`OSIndexAPIImpl.java:512`) | Mirrors ES retry (`OS_CONNECTION_ATTEMPTS` / `ES_CONNECTION_ATTEMPTS`, configurable sleep); on exhaustion → `immediateExit(1)`. **Always aborts — not phase-aware.** |
| Semantic validator | `IndexStartupValidator.validateIndexingConfig()` via `ContentletIndexAPIImpl.checkAndInitializeIndex()` (`ContentletIndexAPIImpl.java:524-541`) | Single probe, **no retry**. Phase-aware: Phase 3 → `DotRuntimeException`; Phase 0–2 → `haltMigration()` (fallback). |

### Startup ordering problem (empty-DB path)

By servlet `load-on-startup` order:

1. `MainServlet` (=1) calls `APILocator.getESIndexAPI().waitUtilIndexReady()` (`MainServlet.java:113`) — **waits on ES only, hardcoded**, even in Phase 3 where OS is primary — then runs the startup tasks.
2. Inside the tasks: `Task00004LoadStarter → DotCMSInitDb.loadStarterSite → refreshAllContent → ContentletIndexAPIImpl.fullReindexStart → initIndex → bootstrapAndPoint → bootstrapAndPointOS → createContentIndex(IndexTag.OS)`. **OS indices are created here, before any OS connection gate.**
3. `InitServlet` (=8) runs `checkAndInitializeIndex()` → the phase-aware validator — **too late**: indices already exist, so `if (!indexReady())` is false and the gate is effectively skipped.

Scope note: the empty-DB bypass only bites when **DB is empty AND migration is started (Phase ≥ 1)**, because `initIndex()` computes `osNeeded = isMigrationStarted() && !indexReadyOS()` (`ContentletIndexAPIImpl.java:688-689`). In Phase 0 no OS indices are created, so there is no gap. This is exactly the `os-migration` fresh-install scenario.

### Target behavior

| Phase | OS unreachable after N retries | Behavior |
|---|---|---|
| Phase 1 / 2 (shadow) | yes | **Fallback to ES** (`haltMigration()` / phase reset) |
| Phase 3 (OS primary) | yes | **Abort**, same as ES does today — no fallback |

Consistent with [feedback] "OS write failures are fire-and-forget in Phase 1/2 but must propagate in Phase 3 (OS is primary)".

## Acceptance Criteria

**A — Startup readiness gate waits on the primary store, not ES-hardcoded**
- [ ] `MainServlet` startup readiness wait routes through the phase-aware router (`getIndexAPI().waitUtilIndexReady()`, `IndexAPIImpl.java:193`) instead of `getESIndexAPI()` directly, so the gate waits on the **primary** store for the current phase (ES in Phase 0–1, OS in Phase 3).
- [ ] In Phase 0 / Phase 1, ES remains the store waited on (no behavior change for non-migrating installs).

**B — OS `waitUtilIndexReady()` outcome is phase-aware**
- [ ] In **Phase 3**, exhausting `OS_CONNECTION_ATTEMPTS` retries aborts startup (current `immediateExit` behavior preserved) with an actionable fatal message.
- [ ] In **Phase 1 / 2**, exhausting retries does **not** kill the server: it falls back to ES via `haltMigration()` (phase reset), logging a clear WARN/ERROR explaining the fallback.
- [ ] Retry count and sleep remain configurable (`OS_CONNECTION_ATTEMPTS` w/ `ES_CONNECTION_ATTEMPTS` fallback; `OS_CONNECTION_RETRY_SLEEP_SECONDS`).

**C — Connection gate runs before OS index creation on the empty-DB path (unification, minimal)**
- [ ] Before `bootstrapAndPointOS` creates OS indices (`ContentletIndexAPIImpl.initIndex`/`bootstrapAndPoint`, ~`682-734`), the primary-store connection gate (retry-N, phase-aware outcome from B) runs first.
- [ ] On empty-DB + Phase ≥ 1, an unreachable/misconfigured OS fails **fast with an actionable message** (config + phase + endpoint), rather than throwing `ConnectionClosedException` deep inside `createContentIndex`.
- [ ] Both startup paths (populated-DB via `InitServlet` and empty-DB via `Task00004LoadStarter`) pass through the same connection gate before creating/using OS indices — the gate no longer lives only in `InitServlet` (which runs too late).
- [ ] **Minimal scope:** C reuses the retry + phase-aware outcome from B. It does **not** move the full semantic `validateIndexingConfig()` (OS version + endpoint-separation) into the bootstrap hot path — that stays in `checkAndInitializeIndex()`.

**General**
- [ ] No regression to non-migrating (Phase 0) startup on either populated or empty DB.
- [ ] New/changed phase-branching behavior covered by tests (unit/IT as appropriate).

## Out of scope (separate follow-up)

- **Error classification hardening** of the connection `catch`: distinguish TLS-scheme mismatch (`http` vs `https`) vs HTTP 403 (auth) vs connection-refused (unreachable), so the fatal/fallback message names the actual cause. This is the diagnostic improvement that motivated the finding but is orthogonal to the gating/fallback rework here — to be filed as a child issue.

## Priority

Medium — reliability gap; bites in OS-migration startup scenarios (empty DB + Phase ≥ 1, or any OS misconfiguration in Phase 3). Not a default-install (Phase 0) regression.

## Additional Context

- Originated from the 2026-06-12 Phase 3 smoke against the `os-migration` limited-user stack.
- Relevant phase semantics: `IndexConfigHelper.MigrationPhase` (`PHASE_0…PHASE_3`), `isMigrationStarted()`, `isMigrationComplete()`, `haltMigration()`.
- Files: `MainServlet.java:113`, `IndexAPIImpl.java:193`, `OSIndexAPIImpl.java:512`, `ESIndexAPI.java:1150`, `ContentletIndexAPIImpl.java:522-558,682-734,874-891`, `IndexStartupValidator.java`, `IndexConfigHelper.java`.

## QA: FAILED ❌
* Docker image: Local image generated from latest `main`.

dotCMS keeps shutting down and starting up when phase 3 is set. It seems like the shutdown process is somehow calling the startup process over and over again, so the application never really finishes. Here's part of the log showing that:

```
dotcms-1                  | 14:39:46.203  ERROR opensearch.OSIndexAPIImpl - OpenSearch Connection Attempt #24 of 24: Connect to https://opensearch3:9210 [opensearch3/172.18.0.2] failed: Connection refused
dotcms-1                  | 14:39:49.896  ERROR opensearch.OSIndexAPIImpl - Error fetching OS cluster stats: Connect to https://opensearch3:9210 [opensearch3/172.18.0.2] failed: Connection refused
dotcms-1                  | java.net.ConnectException: Connect to https://opensearch3:9210 [opensearch3/172.18.0.2] failed: Connection refused
dotcms-1                  | 	at org.opensearch.client.transport.httpclient5.ApacheHttpClient5Transport.extractAndWrapCause(ApacheHttpClient5Transport.java:1141) ~[?:?]
dotcms-1                  | 	at org.opensearch.client.transport.httpclient5.ApacheHttpClient5Transport.performRequest(ApacheHttpClient5Transport.java:160) ~[?:?]
dotcms-1                  | 	at org.opensearch.client.opensearch.nodes.OpenSearchNodesClientBase.stats(OpenSearchNodesClientBase.java:117) ~[?:?]
dotcms-1                  | 	at org.opensearch.client.opensearch.nodes.OpenSearchNodesClientBase.stats(OpenSearchNodesClientBase.java:134) ~[?:?]
dotcms-1                  | 	at com.dotcms.content.index.opensearch.OSIndexAPIImpl.getClusterStats(OSIndexAPIImpl.java:473) ~[?:?]
dotcms-1                  | 	at com.dotcms.content.index.opensearch.OSIndexAPIImpl$Proxy$_$$_WeldClientProxy.getClusterStats(Unknown Source) ~[?:?]
dotcms-1                  | 	at com.dotcms.content.index.PhaseRouter.read(PhaseRouter.java:189) ~[?:?]
dotcms-1                  | 	at com.dotcms.content.index.IndexAPIImpl.getClusterStats(IndexAPIImpl.java:221) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.checks.cdi.ElasticsearchHealthCheck.testElasticsearchConnectivity(ElasticsearchHealthCheck.java:58) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.checks.cdi.ElasticsearchHealthCheck.lambda$performCheck$0(ElasticsearchHealthCheck.java:35) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.util.HealthCheckBase.measureExecution(HealthCheckBase.java:211) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.checks.cdi.ElasticsearchHealthCheck.performCheck(ElasticsearchHealthCheck.java:33) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.util.HealthCheckBase.check(HealthCheckBase.java:77) ~[?:?]
dotcms-1                  | 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
dotcms-1                  | 	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[?:?]
dotcms-1                  | 	at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:38) ~[?:?]
dotcms-1                  | 	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:106) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.api.HealthCheck$920920712$Proxy$_$$_WeldClientProxy.check(Unknown Source) ~[?:?]
dotcms-1                  | 	at com.dotcms.health.service.HealthStateManager.lambda$runSingleHealthCheck$0(HealthStateManager.java:1024) ~[?:?]
dotcms-1                  | 	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1789) ~[?:?]
dotcms-1                  | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545) ~[?:?]
dotcms-1                  | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328) ~[?:?]
dotcms-1                  | 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:309) ~[?:?]
dotcms-1                  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090) ~[?:?]
dotcms-1                  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614) ~[?:?]
dotcms-1                  | 	at java.base/java.lang.Thread.run(Thread.java:1474) [?:?]
dotcms-1                  | Caused by: org.apache.hc.client5.http.HttpHostConnectException: Connect to https://opensearch3:9210 [opensearch3/172.18.0.2] failed: Connection refused
dotcms-1                  | 	at java.base/sun.nio.ch.Net.pollConnect(Native Method) ~[?:?]
dotcms-1                  | 	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:639) ~[?:?]
dotcms-1                  | 	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:1046) ~[?:?]
dotcms-1                  | 	at org.apache.hc.core5.reactor.InternalConnectChannel.onIOEvent(InternalConnectChannel.java:70) ~[?:?]
dotcms-1                  | 	at org.apache.hc.core5.reactor.InternalChannel.handleIOEvent(InternalChannel.java:51) ~[?:?]
dotcms-1                  | 	at org.apache.hc.core5.reactor.SingleCoreIOReactor.processEvents(SingleCoreIOReactor.java:176) ~[?:?]
dotcms-1                  | 	at org.apache.hc.core5.reactor.SingleCoreIOReactor.doExecute(SingleCoreIOReactor.java:125) ~[?:?]
dotcms-1                  | 	at org.apache.hc.core5.reactor.AbstractSingleCoreIOReactor.execute(AbstractSingleCoreIOReactor.java:92) ~[?:?]
dotcms-1                  | 	at org.apache.hc.core5.reactor.IOReactorWorker.run(IOReactorWorker.java:44) ~[?:?]
dotcms-1                  | 	... 1 more
dotcms-1                  | 14:40:18.032  FATAL opensearch.OSIndexAPIImpl - OpenSearch is not reachable after 24 attempt(s). phase=PHASE_3_OPENSEARCH_ONLY, endpoints=[https://opensearch3:9210], cause=Connect to https://opensearch3:9210 [opensearch3/172.18.0.2] failed: Connection refused — OS is the primary store in PHASE_3_OPENSEARCH_ONLY; cannot fall back to ES. Verify OS_ENDPOINTS, OS_PROTOCOL/OS_TLS_ENABLED (scheme must match the server), and credentials, then restart dotCMS.
dotcms-1                  | 14:40:18.034  INFO  shutdown.SystemExitManager - System exit requested: OpenSearch connection failed in PHASE_3_OPENSEARCH_ONLY (exit code: 1)
dotcms-1                  | 14:40:18.034  INFO  shutdown.SystemExitManager - Forcing JVM halt with exit code: 1
dotcms-1 exited with code 1 (restarting)
dotcms-1                  | [FIPS Detection] APR SSL Engine enabled (default) for optimal performance
dotcms-1                  | [FIPS Detection] To disable APR SSL Engine, set CMS_DISABLE_APR_SSL=true or CMS_SSL_ENGINE=off
dotcms-1                  | [FIPS Detection] Final CMS_SSL_ENGINE value: on Detach
dotcms-1                  | SSL certificate already exists at /data/shared/assets/certs/local.dotcms.site.pem
dotcms-1                  | Using default starter
dotcms-1                  | Dump file not found []
dotcms-1                  |
dotcms-1                  | Starting dotCMS ...
dotcms-1                  | -------------------
dotcms-1                  |
dotcms-1                  | setenv: auto-GC: -XX:+UseZGC (18 vCPU; override via JAVA_OPTS_MEMORY)
dotcms-1                  | Setting log4j2.configurationFile=/srv/dotserver/tomcat/webapps/ROOT/WEB-INF/log4j/log4j2.xml
dotcms-1                  | Setting Log4jContextSelector to org.apache.logging.log4j.core.async.BasicAsyncLoggerContextSelector
dotcms-1                  | Disabling log4j automatic shutdown hook - dotCMS will control logging shutdown
dotcms-1                  | Cleaning up temp directory (/srv/dotserver/tomcat/temp) before dotCMS startup...
dotcms-1                  | Removing files older than 60 minutes...
dotcms-1                  | Temp directory cleanup completed
dotcms-1                  | Adding byte-buddy-agent: /srv/dotserver/tomcat/webapps/ROOT/WEB-INF/lib/byte-buddy-agent-1.18.1.jar
dotcms-1                  | Adding Glowroot agent to CATALINA_OPTS
dotcms-1                  | NOTE: Picked up JDK_JAVA_OPTIONS:  --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
dotcms-1                  | OpenJDK 64-Bit Server VM warning: Option -Xdebug was deprecated in JDK 22 and will likely be removed in a future release.
dotcms-1                  | Listening for transport dt_socket at address: 8000
dotcms-1                  | WARNING: A terminally deprecated method in sun.misc.Unsafe has been called
dotcms-1                  | WARNING: sun.misc.Unsafe::arrayBaseOffset has been called by com.lmax.disruptor.RingBufferFields (file:/srv/dotserver/tomcat-9.0.118/log4j2/lib/disruptor-3.3.4.jar)
dotcms-1                  | WARNING: Please consider reporting this to the maintainers of class com.lmax.disruptor.RingBufferFields
dotcms-1                  | WARNING: sun.misc.Unsafe::arrayBaseOffset will be removed in a future release
dotcms-1                  | 14:40:22.638  WARN  net.SSLHostConfigCertificate - The property [Certificate.certificateKeystorePassword] was set on the SSLHostConfigCertificate named [_default_] and is for certificate storage type [KEYSTORE] but the certificate is being used with a storage of type [PEM]
dotcms-1                  | 14:40:22.644  WARN  digester.Digester - Match [Server/Service/Engine/Host/Valve] failed to set property [remoteIpPortHeader] to [x-forwarded-port]
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Server version name:   Apache Tomcat/9.0.118
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Server built:          May 1 2026 19:30:30 UTC
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Server version number: 9.0.118.0
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - OS Name:               Linux
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - OS Version:            6.12.76-linuxkit
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Architecture:          aarch64
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Java Home:             /java
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - JVM Version:           25.0.2+10-LTS
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - JVM Vendor:            Microsoft
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - CATALINA_BASE:         /srv/dotserver/tomcat-9.0.118
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - CATALINA_HOME:         /srv/dotserver/tomcat-9.0.118
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.lang=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.lang.invoke=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.lang.reflect=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.io=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.util=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.util.concurrent=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djava.util.logging.config.file=/srv/dotserver/tomcat/conf/logging.properties
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djava.awt.headless=true
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djava.library.path=/usr/lib/aarch64-linux-gnu/
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -XX:+UseCompactObjectHeaders
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: --enable-preview
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -XX:MaxRAMPercentage=72.0
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -XX:MinHeapFreeRatio=10
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -XX:MaxHeapFreeRatio=50
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djdk.nio.maxCachedBufferSize=262144
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -XX:+UseZGC
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Xdebug
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=*:8000
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djdk.tls.ephemeralDHKeySize=2048
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Djava.protocol.handler.pkgs=org.apache.catalina.webresources
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Dsun.io.useCanonCaches=false
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Dorg.apache.catalina.security.SecurityListener.UMASK=0027
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Dorg.apache.tomcat.util.digester.PROPERTY_SOURCE=org.apache.tomcat.util.digester.EnvironmentPropertySource
dotcms-1                  | 14:40:22.649  INFO  startup.VersionLoggerListener - Command line argument: -Dfile.encoding=UTF8
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --enable-native-access=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.lang=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.io=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.management/com.sun.jmx.mbeanserver=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.lang.ref=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.lang.reflect=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.net=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.nio=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.nio.charset=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.time=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.time.zone=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.util=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.util.concurrent=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.util.concurrent.locks=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/java.util.regex=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/javax.crypto.spec=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.management/javax.management=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/sun.nio.cs=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/sun.util.calendar=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/sun.util.locale=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: --add-opens=java.base/jdk.internal.misc=ALL-UNNAMED
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Djavax.xml.transform.TransformerFactory=com.sun.org.apache.xalan.internal.xsltc.trax.TransformerFactoryImpl
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Djavax.xml.parsers.DocumentBuilderFactory=com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderFactoryImpl
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Djavax.xml.parsers.SAXParserFactory=com.sun.org.apache.xerces.internal.jaxp.SAXParserFactoryImpl
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dlog4j2.configurationFile=/srv/dotserver/tomcat/webapps/ROOT/WEB-INF/log4j/log4j2.xml
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -DLog4jContextSelector=org.apache.logging.log4j.core.async.BasicAsyncLoggerContextSelector
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dlog4j2.shutdownHookEnabled=false
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Djdk.attach.allowAttachSelf=true
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -javaagent:/srv/dotserver/tomcat/webapps/ROOT/WEB-INF/lib/byte-buddy-agent-1.18.1.jar
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -javaagent:/srv/dotserver/tomcat/glowroot/glowroot.jar
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dglowroot.conf.dir=/srv/dotserver/tomcat/glowroot/local-web
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dglowroot.tmp.dir=/tmp
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dglowroot.data.dir=/data/shared/glowroot/data
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dignore.endorsed.dirs=
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dcatalina.base=/srv/dotserver/tomcat
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Dcatalina.home=/srv/dotserver/tomcat
dotcms-1                  | 14:40:22.650  INFO  startup.VersionLoggerListener - Command line argument: -Djava.io.tmpdir=/srv/dotserver/tomcat/temp
dotcms-1                  | 14:40:22.652  ERROR core.AprLifecycleListener - An incompatible version [1.2.35] of the Apache Tomcat Native library is installed, while Tomcat requires version [1.3.4]
dotcms-1                  | 14:40:22.768  INFO  http11.Http11Nio2Protocol - Initializing ProtocolHandler ["http-nio2-8080"]
dotcms-1                  | 14:40:22.772  WARN  net.Nio2Endpoint - The NIO2 connector requires an exclusive executor to operate properly on shutdown
dotcms-1                  | 14:40:22.776  INFO  http11.Http11Nio2Protocol - Initializing ProtocolHandler ["http-nio2-8081"]
dotcms-1                  | 14:40:22.777  WARN  net.Nio2Endpoint - The NIO2 connector requires an exclusive executor to operate properly on shutdown
dotcms-1                  | 14:40:22.777  INFO  http11.Http11Nio2Protocol - Initializing ProtocolHandler ["http-nio2-8082"]
dotcms-1                  | 14:40:22.777  WARN  net.Nio2Endpoint - The NIO2 connector requires an exclusive executor to operate properly on shutdown
dotcms-1                  | 14:40:22.783  INFO  http11.Http11Nio2Protocol - Initializing ProtocolHandler ["https-jsse-nio2-8443"]
dotcms-1                  | 14:40:22.783  WARN  net.Nio2Endpoint - The NIO2 connector requires an exclusive executor to operate properly on shutdown
dotcms-1                  | 14:40:22.877  INFO  Nio2Endpoint.certificate - Connector [https-jsse-nio2-8443], TLS virtual host [_default_], certificate type [UNDEFINED] configured from key [/data/shared/assets/certs/local.dotcms.site-key.pem], certificate [/data/shared/assets/certs/local.dotcms.site.pem] and certificate chain [null] with trust store [null]
dotcms-1                  | 14:40:22.879  INFO  http11.Http11Nio2Protocol - Initializing ProtocolHandler ["http-nio2-0.0.0.0-8090"]
dotcms-1                  | 14:40:22.884  INFO  startup.Catalina - Server initialization in [319] milliseconds
dotcms-1                  | 14:40:22.892  INFO  core.StandardService - Starting service [Catalina]
dotcms-1                  | 14:40:22.892  INFO  core.StandardEngine - Starting Servlet engine: [Apache Tomcat/9.0.118]
dotcms-1                  | 14:40:22.898  INFO  startup.HostConfig - Deploying web application directory [/srv/dotserver/tomcat-9.0.118/webapps/ROOT]
dotcms-1                  | 14:40:27.188  INFO  util.Config - Initializing properties reader.
dotcms-1                  | 14:40:27.189  INFO  util.Config - Loading dotCMS [dotmarketing-config.properties] Properties...
dotcms-1                  | 14:40:27.208  INFO  util.Config - dotCMS Properties [dotmarketing-config.properties] Loaded
dotcms-1                  | 14:40:27.223  INFO  util.Config - Loading dotCMS [dotcms-config-cluster.properties] Properties...
dotcms-1                  | 14:40:27.223  INFO  util.Config - dotCMS Properties [dotcms-config-cluster.properties] Loaded
dotcms-1                  | 14:40:27.321  INFO  servlet.TldScanner - At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
dotcms-1                  | 14:40:27.339  INFO  environment.servletWeldServlet - WELD-ENV-001008: Initialize Weld using ServletContainerInitializer
dotcms-1                  | 14:40:27.350  INFO  weld.Version - WELD-000900: 3.1.9 (Final)
dotcms-1                  | 14:40:27.429  INFO  support.LettuceCdiExtension - Activating CDI extension for lettuce.
dotcms-1                  | 14:40:27.437  INFO  weld.Bootstrap - WELD-ENV-000020: Using jandex for bean discovery
dotcms-1                  | 14:40:28.223  INFO  weld.Bootstrap - WELD-000101: Transactional services not available. Injection of @Inject UserTransaction not available. Transactional observers will be invoked synchronously.
dotcms-1                  | 14:40:28.289  INFO  weld.Event - WELD-000411: Observer method [BackedAnnotatedMethod] public org.glassfish.jersey.ext.cdi1x.internal.ProcessAllAnnotatedTypes.processAnnotatedType(@Observes ProcessAnnotatedType<?>, BeanManager) receives events for all annotated types. Consider restricting events using @WithAnnotations or a generic type with bounds.
dotcms-1                  | 14:40:28.385  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.EnterpriseFeatureInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.386  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.CloseDBIfOpenedInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.386  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.CloseDBInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.386  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.WrapInTransactionInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.386  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.ExternalTransactionInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.386  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.LogTimeInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.386  WARN  weld.Validator - WELD-001478: Interceptor class com.dotcms.cdi.business.RequestCostInterceptor is enabled for the application and for the bean archive _/WEB-INF/classes. It will only be invoked in the @Priority part of the chain.
dotcms-1                  | 14:40:28.392  INFO  environment.servletTomcat - WELD-ENV-001100: Tomcat 7+ detected, CDI injection will be available in Servlets, Filters and Listeners.
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -            OOOO                            7777777   7777       7777     77777777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -            OOOO                 OO       777777777  77777      77777    77777777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -            OOOO                OOO      77777       777777     77777   7777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -      OOOOOOOOOO   OOOOOOOO   OOOOOOOOO 7777         7777777   777777   7777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -     OOOO  OOOOO  OOOO  OOOO    OOOO    7777         7777777  7777777    777777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -    OOOO    OOOO  OOO    OOOO   OOOO    7777         77777777 777 7777     777777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -    OOOO    OOOO OOOO    OOOO   OOOO    7777         777  777 777 7777        7777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -    OOOO    OOOO  OOO    OOOO   OOOO    77777        777  777777  7777         7777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -     OOOO   OOOO  OOOO   OOO    OOOO     77777       777   7777   7777        7777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -      OOOOOOOOOO   OOOOOOOO      OOOOO    777777777  777   7777   7777  777777777
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -                                                          Content Management System
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -
dotcms-1                  | 14:40:28.859  INFO  util.AsciiArt -
dotcms-1                  | 14:40:29.918  INFO  cost.LeakyTokenBucketImpl$Proxy$_$$_WeldClientProxy - Rate limiting enabled: false, refill per second: 500, max bucket size: 10000
dotcms-1                  | 14:40:29.922  INFO  environment.servletWeldServlet - WELD-ENV-001006: org.jboss.weld.environment.servlet.EnhancedListener used to initialize Weld
dotcms-1                  | 14:40:29.922  INFO  listeners.ContextLifecycleListener - ContextLifecycleListener contextInitialized called
...
...
...
```

--- a/dotCMS/src/main/java/com/dotcms/content/index/opensearch/OSIndexAPIImpl.java
+++ b/dotCMS/src/main/java/com/dotcms/content/index/opensearch/OSIndexAPIImpl.java
@@ -36,6 +36,7 @@ import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
@@ -89,6 +90,23 @@ public class OSIndexAPIImpl implements IndexAPI {
     private OSClientProvider clientProvider;
 
     /**
+     * Seam over the JVM-terminating exit used by the Phase 3 abort branch of the startup
+     * connection gate. Defaults to the real {@link com.dotcms.shutdown.SystemExitManager}, which
+     * calls {@code Runtime.halt()} and has no test guard. Tests override this via
+     * {@link #setExitHandler(ExitHandler)} so the Phase 3 abort contract can be asserted without
+     * killing the test JVM (issue #36244 follow-up).
+     */
+    private ExitHandler exitHandler = com.dotcms.shutdown.SystemExitManager::immediateExit;
+
+    /**
+     * Functional seam for {@link com.dotcms.shutdown.SystemExitManager#immediateExit(int, String)}.
+     */
+    @FunctionalInterface
+    interface ExitHandler {
+        void exit(int exitCode, String reason);
+    }
+
+    /**
      * No-arg constructor required by CDI for proxy creation.
      * The {@code clientProvider} dependency is injected via field injection after construction.
      */
@@ -101,6 +119,14 @@ public class OSIndexAPIImpl implements IndexAPI {
      */
     OSIndexAPIImpl(OSClientProvider clientProvider) {
         this.clientProvider = clientProvider;
+    }
+
+    /**
+     * Package-private test seam: replaces the JVM-terminating exit so the Phase 3 abort branch of
+     * {@link #waitUtilIndexReady()} can be observed in tests instead of halting the JVM.
+     */
+    void setExitHandler(final ExitHandler exitHandler) {
+        this.exitHandler = exitHandler;
     }
 
     // =========================================================================
@@ -573,20 +599,21 @@ public class OSIndexAPIImpl implements IndexAPI {
     private boolean handleConnectionExhausted(final int attempts, final Exception lastError) {
         final MigrationPhase phase = MigrationPhase.current();
         final String cause = lastError != null ? lastError.getMessage() : "unknown";
+        final ConnectionFailureKind kind = classifyConnectionError(lastError);
         final String detail = "OpenSearch is not reachable after " + attempts + " attempt(s)."
                 + " phase=" + phase.name()
                 + ", endpoints=" + resolveEndpointsForLogging()
-                + ", cause=" + cause;
+                + ", likelyCause=" + kind.name() + " (" + kind.remediation() + ")"
+                + ", error=" + cause;
 
         if (phase.isMigrationComplete()) {
             // Phase 3: OS is primary and ES is decommissioned — no fallback is possible.
             Logger.fatal(this.getClass(), detail
                     + " — OS is the primary store in " + phase.name() + "; cannot fall back to ES."
-                    + " Verify OS_ENDPOINTS, OS_PROTOCOL/OS_TLS_ENABLED (scheme must match the"
-                    + " server), and credentials, then restart dotCMS.");
-            com.dotcms.shutdown.SystemExitManager.immediateExit(1,
-                    "OpenSearch connection failed in PHASE_3_OPENSEARCH_ONLY");
-            return false; // unreachable — immediateExit terminates the JVM
+                    + " Fix the connection per the likely cause above, then restart dotCMS.");
+            exitHandler.exit(1, "OpenSearch connection failed in PHASE_3_OPENSEARCH_ONLY: "
+                    + kind.name());
+            return false; // unreachable in production — immediateExit terminates the JVM
         }
 
         // Phase 1 / 2 (shadow): ES still holds the authoritative state. Fall back to ES-only
@@ -594,7 +621,8 @@ public class OSIndexAPIImpl implements IndexAPI {
         Logger.error(this.getClass(), detail
                 + " — OS is a shadow store in " + phase.name() + "; falling back to ES-only"
                 + " (resetting FEATURE_FLAG_OPEN_SEARCH_PHASE to 0 via haltMigration)."
-                + " Fix OS connectivity and re-enable the migration phase when ready.");
+                + " Fix OS connectivity per the likely cause above and re-enable the migration"
+                + " phase when ready.");
         IndexConfigHelper.haltMigration();
         return false;
     }
@@ -614,6 +642,87 @@ public class OSIndexAPIImpl implements IndexAPI {
         final String hostname = IndexConfigHelper.getString(OSIndexProperty.HOSTNAME, "localhost");
         final int    port     = IndexConfigHelper.getInt(OSIndexProperty.PORT, 9200);
         return protocol + "://" + hostname + ":" + port;
+    }
+
+    /**
+     * Likely root cause of an OpenSearch connection failure, inferred from the exception chain,
+     * so the fatal (Phase 3) / fallback (Phase 1-2) message can name the actual problem and its fix
+     * instead of only echoing a raw transport message (issue #36244 follow-up: error classification
+     * hardening). Motivated by a TLS-scheme mismatch (an {@code http://} client against an
+     * {@code https}-only OS 3.x port) that surfaced only as an opaque {@code ConnectionClosedException}.
+     */
+    enum ConnectionFailureKind {
+        TLS_SCHEME_MISMATCH("TLS/scheme mismatch — the client scheme likely does not match the"
+                + " server (e.g. http:// against an https-only OS port, or https:// against a"
+                + " plaintext port). Align OS_PROTOCOL / OS_TLS_ENABLED with the server's scheme"
+                + " and check the server certificate."),
+        AUTH_FORBIDDEN("authentication/authorization rejected (HTTP 401/403) — verify the OS"
+                + " credentials and that the user has the required cluster permissions."),
+        UNREACHABLE("cluster unreachable (connection refused / timed out / unknown host) — verify"
+                + " OS_ENDPOINTS host:port and that OpenSearch is running and network-reachable."),
+        UNKNOWN("cause could not be classified — inspect the underlying error and stack trace.");
+
+        private final String remediation;
+
+        ConnectionFailureKind(final String remediation) {
+            this.remediation = remediation;
+        }
+
+        String remediation() {
+            return remediation;
+        }
+    }
+
+    /**
+     * Classifies an OpenSearch connection failure by walking the exception chain and matching on
+     * exception type (simple name) and message text. Simple-name / message matching is used
+     * deliberately so this stays free of any compile dependency on transport-layer types (e.g.
+     * Apache HttpClient's {@code ConnectionClosedException}) and remains unit-testable with plain
+     * synthetic exception chains.
+     *
+     * <p>Order matters: TLS/scheme is checked first (its symptoms — closed connection, SSL/plaintext
+     * — otherwise look like a generic unreachable), then auth (401/403), then unreachable.</p>
+     *
+     * @param error the last connection error observed, or {@code null}
+     * @return the inferred {@link ConnectionFailureKind}; never {@code null}
+     */
+    static ConnectionFailureKind classifyConnectionError(final Throwable error) {
+        for (Throwable t = error; t != null; t = t.getCause()) {
+            final String type = t.getClass().getName().toLowerCase(Locale.ROOT);
+            final String msg  = t.getMessage() == null
+                    ? "" : t.getMessage().toLowerCase(Locale.ROOT);
+
+            // TLS / scheme mismatch: SSL handshake errors, or a connection closed with no response
+            // (the classic symptom of speaking http:// to an https-only port).
+            if (type.contains("sslexception")
+                    || type.contains("sslhandshake")
+                    || type.contains("connectionclosedexception")
+                    || msg.contains("unrecognized ssl message")
+                    || msg.contains("plaintext")
+                    || msg.contains("ssl")
+                    || msg.contains("certificat") // certificate / certification path (PKIX)
+                    || msg.contains("pkix")
+                    || msg.contains("connection closed")) {
+                return ConnectionFailureKind.TLS_SCHEME_MISMATCH;
+            }
+            // Authentication / authorization.
+            if (msg.contains("401") || msg.contains("403")
+                    || msg.contains("unauthorized") || msg.contains("forbidden")) {
+                return ConnectionFailureKind.AUTH_FORBIDDEN;
+            }
+            // Host/port unreachable.
+            if (type.contains("connectexception")
+                    || type.contains("unknownhostexception")
+                    || type.contains("sockettimeoutexception")
+                    || type.contains("nohttpresponseexception")
+                    || msg.contains("connection refused")
+                    || msg.contains("timed out")
+                    || msg.contains("timeout")
+                    || msg.contains("unknown host")) {
+                return ConnectionFailureKind.UNREACHABLE;
+            }
+        }
+        return ConnectionFailureKind.UNKNOWN;
     }
 
 

--- a/dotCMS/src/test/java/com/dotcms/content/index/opensearch/OSIndexAPIImplConnectionClassifyTest.java
+++ b/dotCMS/src/test/java/com/dotcms/content/index/opensearch/OSIndexAPIImplConnectionClassifyTest.java
@@ -1,0 +1,111 @@
+package com.dotcms.content.index.opensearch;
+
+import static org.junit.Assert.assertEquals;
+
+import com.dotcms.content.index.opensearch.OSIndexAPIImpl.ConnectionFailureKind;
+import java.io.IOException;
+import java.net.ConnectException;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import javax.net.ssl.SSLException;
+import org.junit.Test;
+
+/**
+ * Unit tests for {@link OSIndexAPIImpl#classifyConnectionError(Throwable)} — the OpenSearch
+ * connection-error classification hardening (issue #36244 follow-up).
+ *
+ * <p>Pure function, no container: the classifier walks the exception chain and matches on exception
+ * simple-name and message text, so synthetic exception chains are sufficient to exercise every
+ * branch. Same package as {@link OSIndexAPIImpl} to reach the package-private classifier and
+ * {@link ConnectionFailureKind}.</p>
+ *
+ * @author Fabrizzio Araya
+ */
+public class OSIndexAPIImplConnectionClassifyTest {
+
+    /** Synthetic type whose simple name matches Apache HttpClient's transport exception. */
+    private static final class ConnectionClosedException extends IOException {
+        ConnectionClosedException(final String message) {
+            super(message);
+        }
+    }
+
+    private static ConnectionFailureKind classify(final Throwable t) {
+        return OSIndexAPIImpl.classifyConnectionError(t);
+    }
+
+    // ---- TLS / scheme mismatch ------------------------------------------------------------------
+
+    @Test
+    public void sslException_isTlsSchemeMismatch() {
+        assertEquals(ConnectionFailureKind.TLS_SCHEME_MISMATCH,
+                classify(new SSLException("Unrecognized SSL message, plaintext connection?")));
+    }
+
+    @Test
+    public void connectionClosed_httpAgainstHttpsPort_isTlsSchemeMismatch() {
+        // The classic symptom of speaking http:// to an https-only OS 3.x port.
+        assertEquals(ConnectionFailureKind.TLS_SCHEME_MISMATCH,
+                classify(new ConnectionClosedException("Connection is closed")));
+    }
+
+    @Test
+    public void certificateMessage_isTlsSchemeMismatch() {
+        assertEquals(ConnectionFailureKind.TLS_SCHEME_MISMATCH,
+                classify(new IOException("unable to find valid certification path to requested target")));
+    }
+
+    // ---- Auth -----------------------------------------------------------------------------------
+
+    @Test
+    public void forbidden403_isAuthForbidden() {
+        assertEquals(ConnectionFailureKind.AUTH_FORBIDDEN,
+                classify(new RuntimeException("method [HEAD], status line [HTTP/1.1 403 Forbidden]")));
+    }
+
+    @Test
+    public void unauthorized401_isAuthForbidden() {
+        assertEquals(ConnectionFailureKind.AUTH_FORBIDDEN,
+                classify(new RuntimeException("401 Unauthorized")));
+    }
+
+    // ---- Unreachable ----------------------------------------------------------------------------
+
+    @Test
+    public void connectException_isUnreachable() {
+        assertEquals(ConnectionFailureKind.UNREACHABLE,
+                classify(new ConnectException("Connection refused")));
+    }
+
+    @Test
+    public void unknownHost_isUnreachable() {
+        assertEquals(ConnectionFailureKind.UNREACHABLE,
+                classify(new UnknownHostException("no-such-host")));
+    }
+
+    @Test
+    public void socketTimeout_isUnreachable() {
+        assertEquals(ConnectionFailureKind.UNREACHABLE,
+                classify(new SocketTimeoutException("connect timed out")));
+    }
+
+    // ---- Cause-chain walking & fallback ---------------------------------------------------------
+
+    @Test
+    public void walksCauseChain_wrappedSslException_isTlsSchemeMismatch() {
+        final Throwable wrapped = new RuntimeException("io error",
+                new IOException("handshake failed", new SSLException("bad_certificate")));
+        assertEquals(ConnectionFailureKind.TLS_SCHEME_MISMATCH, classify(wrapped));
+    }
+
+    @Test
+    public void nullError_isUnknown() {
+        assertEquals(ConnectionFailureKind.UNKNOWN, classify(null));
+    }
+
+    @Test
+    public void unrecognizedError_isUnknown() {
+        assertEquals(ConnectionFailureKind.UNKNOWN,
+                classify(new IllegalStateException("something else entirely")));
+    }
+}

--- a/dotcms-integration/src/test/java/com/dotcms/content/index/opensearch/OSIndexAPIImplWaitReadyIT.java
+++ b/dotcms-integration/src/test/java/com/dotcms/content/index/opensearch/OSIndexAPIImplWaitReadyIT.java
@@ -3,10 +3,13 @@ package com.dotcms.content.index.opensearch;
 import static com.dotcms.content.index.IndexConfigHelper.MigrationPhase.FLAG_KEY;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
 
 import com.dotcms.content.index.IndexConfigHelper.MigrationPhase;
 import com.dotcms.util.IntegrationTestInitService;
 import com.dotmarketing.util.Config;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.BeforeClass;
@@ -35,9 +38,12 @@ import org.opensearch.client.opensearch.OpenSearchClient;
  *   <li><strong>Phase 1 / 2 (shadow)</strong> — covered: the gate must NOT kill the JVM; it halts
  *       the migration (phase reset to 0) and returns {@code false}. These two cases use the failing
  *       client provider, so they do not need the live OpenSearch container.</li>
- *   <li><strong>Phase 3 (OS primary)</strong> — intentionally NOT tested here: the gate aborts via
- *       {@code SystemExitManager.immediateExit}, which calls {@code Runtime.halt()} and would kill
- *       the test JVM. The Phase 3 abort is verified by manual QA instead.</li>
+ *   <li><strong>Phase 3 (OS primary)</strong> — covered via a test exit seam
+ *       ({@link OSIndexAPIImpl#setExitHandler}): the gate must request an abort (exit code 1) and
+ *       must NOT fall back (phase stays {@code PHASE_3}). The seam replaces
+ *       {@code SystemExitManager.immediateExit} — which calls {@code Runtime.halt()} and would
+ *       otherwise kill the test JVM — so the abort contract is observable without terminating the
+ *       process.</li>
  *   <li><strong>Success path</strong> — covered by the other OpenSearch integration tests that run
  *       against the live cluster (a reachable {@code client.info()}).</li>
  * </ul>
@@ -111,5 +117,61 @@ public class OSIndexAPIImplWaitReadyIT {
         assertFalse("Phase 2 must NOT abort: the gate returns false (ES-only fallback)", ready);
         assertEquals("Migration phase must be reset to PHASE_0 after the ES fallback",
                 MigrationPhase.PHASE_0_MIGRATION_NOT_STARTED, MigrationPhase.current());
+    }
+
+    /**
+     * Given : Phase 3 (OpenSearch only) and an unreachable OpenSearch cluster.
+     * When  : waitUtilIndexReady() exhausts its retries.
+     * Then  : the gate ABORTS (requests exit code 1 via the seam) and does NOT fall back — OS is the
+     *         primary store in Phase 3, so the migration phase must remain PHASE_3 (haltMigration is
+     *         never called). The exit is captured by the seam instead of killing the test JVM.
+     */
+    @Test
+    public void test_phase3_osUnreachable_aborts_noFallback() {
+        setPhase(3);
+        final OSIndexAPIImpl api = new OSIndexAPIImpl(new FailingClientProvider());
+
+        final AtomicInteger exitCode = new AtomicInteger(Integer.MIN_VALUE);
+        final AtomicReference<String> exitReason = new AtomicReference<>();
+        api.setExitHandler((code, reason) -> {
+            exitCode.set(code);
+            exitReason.set(reason);
+        });
+
+        final boolean ready = api.waitUtilIndexReady();
+
+        assertFalse("Phase 3 gate must not report OS as ready when unreachable", ready);
+        assertEquals("Phase 3 must request a JVM abort with exit code 1", 1, exitCode.get());
+        assertTrue("Abort reason must identify the Phase 3 store",
+                exitReason.get() != null && exitReason.get().contains("PHASE_3"));
+        assertEquals("Phase 3 must NOT fall back: the migration phase must remain PHASE_3",
+                MigrationPhase.PHASE_3_OPENSEARCH_ONLY, MigrationPhase.current());
+    }
+
+    /**
+     * Given : Phase 3 and an unreachable cluster whose failure looks like a TLS/scheme mismatch.
+     * When  : waitUtilIndexReady() exhausts its retries.
+     * Then  : the abort reason names the classified cause (TLS_SCHEME_MISMATCH), proving the
+     *         error-classification hardening reaches the actionable Phase 3 message.
+     */
+    @Test
+    public void test_phase3_tlsMismatch_abortReasonNamesCause() {
+        setPhase(3);
+        final OSIndexAPIImpl api = new OSIndexAPIImpl(new OSClientProvider() {
+            @Override
+            public OpenSearchClient getClient() {
+                throw new RuntimeException(
+                        new javax.net.ssl.SSLException("Unrecognized SSL message, plaintext connection?"));
+            }
+        });
+
+        final AtomicReference<String> exitReason = new AtomicReference<>();
+        api.setExitHandler((code, reason) -> exitReason.set(reason));
+
+        api.waitUtilIndexReady();
+
+        assertTrue("Abort reason must name the classified TLS/scheme cause",
+                exitReason.get() != null && exitReason.get().contains(
+                        OSIndexAPIImpl.ConnectionFailureKind.TLS_SCHEME_MISMATCH.name()));
     }
 }

